### ISSUE TITLE
Conversion between WGS84 (lon-lat) and Web Mercator Quad tile matrix set

### ISSUE THREAD

**Issue by navispatial** [navibyte/geospatial#57]
See #51 and #55 for background.

Define and implement also projection from WGS84 (lon-lat) to Web Mercator projection in map pixel coordinates.

**Comment by navispatial** [comment]
`TileMatrixSet` definition:

```dart
/// A tiling scheme represented as a set of tile matrices (grids).
abstract class TileMatrixSet {
  /// The maximum (suggested) zoom level for this tile matrix set.
  ///
  /// Note that some methods may allow larger zoom values when zoom value is
  /// given as an argument.
  int get maxZoom;

  /// The number of pixels for one tile in one axis (normally 256 or 512).
  int get tileSize;

  /// The position of the origin in a tile matrix and map pixel "canvas".
  TileMatrixOrigin get origin;

  /// The number of tiles ("matrix width") at [zoom] in the X axis.
  int matrixWidth(int zoom);

  /// The number of tiles ("matrix height") at [zoom] in the Y axis.
  int matrixHeight(int zoom);

  /// The number of pixels ("map width") at [zoom] in the X axis.
  int mapWidth(int zoom);

  /// The number of pixels ("map height") at [zoom] in the Y axis.
  int mapHeight(int zoom);

  /// The tile ground resolution in meters at [zoom].
  ///
  /// Should equal to `tileSize * pixelResolution(zoom)`.
  ///
  /// This is a nominal resolution that may be accurate only in some positions
  /// like along the equator or a meridian depending on a projection.
  double tileResolution(int zoom);

  /// The pixel ground resolution in meters at [zoom].
  ///
  /// Should equal to `tileResolution(zoom) / tileSize`.
  ///
  /// This is a nominal resolution that may be accurate only in some positions
  /// like along the equator or a meridian depending on a projection.
  double pixelResolution(int zoom);

  /// The map scale denominator at [zoom] and [screenDpi].
  ///
  /// This is a nominal scale denominator that may be accurate only in some
  /// positions like along the equator or a meridian depending on a projection.
  double scaleDenominator(
    int zoom, {
    double screenDpi = 96,
  });

  /// A map converter between geospatial positions and map coordinates.
  MapConverter get converter;

  /// Transforms [position] to world coordinates.
  ///
  /// Coordinate value ranges:
  /// - world x (double): `0.0 .. mapWidth(0)`
  /// - world y (double): `0.0 .. mapHeight(0)`
  Projected positionToWorld(covariant Position position);

  /// Transforms [position] to pixel coordinates at [zoom].
  ///
  /// Coordinate value ranges:
  /// - pixel x (int): `0 .. mapWidth(zoom) - 1`
  /// - pixel y (int): `0 .. mapHeight(zoom) - 1`
  MapPoint2i positionToPixel(covariant Position position, {int zoom = 0});

  /// Returns a tile at [zoom] covering a region in [position].
  ///
  /// Coordinate value ranges:
  /// - tile x (int): `0 .. matrixWidth(zoom) - 1`
  /// - tile y (int): `0 .. matrixHeight(zoom) - 1`
  MapPoint2i positionToTile(covariant Position position, {int zoom = 0});

  /// Transforms [world] coordinates to pixel coordinates at [zoom].
  ///
  /// Coordinate value ranges:
  /// - world x (double): `0.0 .. mapWidth(0)`
  /// - world y (double): `0.0 .. mapHeight(0)`
  /// - pixel x (int): `0 .. mapWidth(zoom) - 1`
  /// - pixel y (int): `0 .. mapHeight(zoom) - 1`
  MapPoint2i worldToPixel(Projected world, {int zoom = 0});

  /// Returns a tile at [zoom] covering a region in [world] coordinates.
  ///
  /// Coordinate value ranges:
  /// - world x (double): `0.0 .. mapWidth(0)`
  /// - world y (double): `0.0 .. mapHeight(0)`
  /// - tile x (int): `0 .. matrixWidth(zoom) - 1`
  /// - tile y (int): `0 .. matrixHeight(zoom) - 1`
  MapPoint2i worldToTile(Projected world, {int zoom = 0});

  /// Transforms [pixel] coordinates to a position.
  ///
  /// Coordinate value ranges:
  /// - pixel x (int): `0 .. mapWidth(zoom) - 1`
  /// - pixel y (int): `0 .. mapHeight(zoom) - 1`
  Position pixelToPosition(MapPoint2i pixel);

  /// Returns a tile covering a region in [pixel] coordinates.
  ///
  /// Coordinate value ranges:
  /// - pixel x (int): `0 .. mapWidth(zoom) - 1`
  /// - pixel y (int): `0 .. mapHeight(zoom) - 1`
  /// - tile x (int): `0 .. matrixWidth(zoom) - 1`
  /// - tile y (int): `0 .. matrixHeight(zoom) - 1`
  MapPoint2i pixelToTile(MapPoint2i pixel);

  /// Returns a bounding box with min and max positions for [tile].
  ///
  /// Coordinate value ranges:
  /// - tile x (int): `0 .. matrixWidth(zoom) - 1`
  /// - tile y (int): `0 .. matrixHeight(zoom) - 1`
  Box tileToBounds(MapPoint2i tile);
}
```

`GeoTileMatrixSet` definition:

```dart
/// A tile matrix set with conversions between tiles and geographic positions.
abstract class GeoTileMatrixSet extends TileMatrixSet {
  /// The tile (longitudal) width in degrees at [zoom].
  double tileWidthLongitudal(int zoom) => 360.0 / matrixWidth(zoom);

  /// The pixel (longitudal) width in degrees at [zoom].
  double pixelWidthLongitudal(int zoom) => 360.0 / mapWidth(zoom);

  @override
  Projected positionToWorld(Geographic position);

  @override
  MapPoint2i positionToPixel(Geographic position, {int zoom = 0});

  @override
  MapPoint2i positionToTile(Geographic position, {int zoom = 0});

  @override
  Geographic pixelToPosition(MapPoint2i pixel);

  @override
  GeoBox tileToBounds(MapPoint2i tile);
```

`WebMercatorQuad` definition:

```dart
/// "Web Mercator Quad" tile matrix set aka "Google Maps Compatible".
///
/// More information:
/// https://en.wikipedia.org/wiki/Web_Mercator_projection
/// https://developers.google.com/maps/documentation/javascript/coordinates
/// https://docs.microsoft.com/en-us/bingmaps/articles/bing-maps-tile-system
@immutable
class WebMercatorQuad extends GeoTileMatrixSet {
  /// Create "Web Mercator Quad" tile matrix set with [tileSize] and [origin].
  ///
  /// Normally use 256 or 512 pixels for [tileSize].
  ///
  /// By default [origin] for tiles and pixels is "top-left", for example
  /// a tile (0, 0) is located in the north-west corner of the world map. This
  /// is the notation used by "Google Maps Compatible" tile services. You can
  /// set [origin] to "bottom-left", the notation used by TMS (Tile Map Service)
  /// specification.
  ///
  /// This implementation is based on the WGS 84 / Web Mercator projection
  /// ("EPSG:3857") aka "Pseudo-Mercator" or "Spherical Mercator".
  WebMercatorQuad.epsg3857({
    this.maxZoom = 22,
    this.tileSize = 256,
    this.origin = TileMatrixOrigin.topLeft,
  });

  /// The number of tiles at [zoom] (level of detail) in one axis.
  int matrixSize(int zoom) => 1 << zoom;

  /// The number of pixels at [zoom] (level of detail) in one axis.
  int mapSize(int zoom) => tileSize << zoom;

  /// The pixel ground resolution in meters at given [latitude] and [zoom].
  double pixelResolutionAt({required double latitude, required int zoom});

  /// The map scale denominator at given [latitude], [zoom] and [screenDpi].
  double scaleDenominatorAt({
    required double latitude,
    required int zoom,
    double screenDpi = 96,
  });

  /// Returns a tile identified by [quadKey] (as specified by Microsoft).
  ///
  /// Throws `FormatException` if [quadKey] is invalid.
  ///
  /// See also:
  /// https://docs.microsoft.com/en-us/bingmaps/articles/bing-maps-tile-system
  MapPoint2i quadKeyToTile(String quadKey);

  /// Returns the quad key (as specified by Microsoft) for [tile].
  ///
  /// See also:
  /// https://docs.microsoft.com/en-us/bingmaps/articles/bing-maps-tile-system
  String tileToQuadKey(MapPoint2i tile);
```

Also other classes needed by previous classes.

**Comment by navispatial** [comment]
Implemented in pre-relase 0.2.0-dev.7 of beta 0.2.0.
